File: rust/src/service_impl.rs
use crate::error::{GoIamError, Result};
use crate::service::Service;
use crate::types::{AuthCallbackResponse, Resource, ResourceResponse, User, UserResponse};
use base64::{engine::general_purpose, Engine as _};
use reqwest::Client;

pub struct ServiceImpl {
    base_url: String,
    client_id: String,
    secret: String,
    client: Client,
}

impl ServiceImpl {
    pub fn new(base_url: String, client_id: String, secret: String) -> Self {
        Self {
            base_url: base_url.trim_end_matches('/').to_string(),
            client_id,
            secret,
            client: Client::new(),
        }
    }

    fn basic_auth(&self) -> String {
        let credentials = format!("{}:{}", self.client_id, self.secret);
        format!("Basic {}", general_purpose::STANDARD.encode(credentials))
    }
}

#[async_trait::async_trait]
impl Service for ServiceImpl {
    async fn verify(&self, code: &str) -> Result<String> {
        let url = format!("{}/auth/v1/verify", self.base_url);

        let response = self
            .client
            .get(&url)
            .query(&[("code", code)])
            .header("Authorization", self.basic_auth())
            .send()
            .await?;

        if !response.status().is_success() {
            return Err(GoIamError::ApiError {
                message: format!("Failed to verify code: {}", response.status()),
                status: response.status().as_u16(),
            });
        }

        let auth_response: AuthCallbackResponse = response.json().await?;

        if !auth_response.success {
            return Err(GoIamError::AuthError {
                message: auth_response
                    .message
                    .unwrap_or_else(|| "Authentication failed".to_string()),
            });
        }

        match auth_response.data {
            Some(data) => Ok(data.access_token),
            None => Err(GoIamError::InvalidResponse {
                message: "No access token received".to_string(),
            }),
        }
    }

    async fn me(&self, token: &str) -> Result<User> {
        let url = format!("{}/me/v1/me", self.base_url);

        let response = self
            .client
            .get(&url)
            .header("Authorization", format!("Bearer {}", token))
            .send()
            .await?;

        if !response.status().is_success() {
            return Err(GoIamError::ApiError {
                message: format!("Failed to fetch user information: {}", response.status()),
                status: response.status().as_u16(),
            });
        }

        let user_response: UserResponse = response.json().await?;

        if !user_response.success {
            return Err(GoIamError::AuthError {
                message: user_response
                    .message
                    .unwrap_or_else(|| "User fetch failed".to_string()),
            });
        }

        match user_response.data {
            Some(user) => Ok(user),
            None => Err(GoIamError::InvalidResponse {
                message: "No user data received".to_string(),
            }),
        }
    }

    async fn create_resource(&self, resource: &Resource, token: &str) -> Result<()> {
        let url = format!("{}/resource/v1/", self.base_url);

        let response = self
            .client
            .post(&url)
            .header("Authorization", format!("Bearer {}", token))
            .header("Content-Type", "application/json")
            .json(resource)
            .send()
            .await?;

        if !response.status().is_success() {
            return Err(GoIamError::ApiError {
                message: format!("Failed to create resource: {}", response.status()),
                status: response.status().as_u16(),
            });
        }

        let resource_response: ResourceResponse = response.json().await?;

        if !resource_response.success {
            return Err(GoIamError::AuthError {
                message: resource_response
                    .message
                    .unwrap_or_else(|| "Resource creation failed".to_string()),
            });
        }

        Ok(())
    }
}

/// Create a new instance of the Go IAM service
pub fn new_service(base_url: String, client_id: String, secret: String) -> impl Service {
    ServiceImpl::new(base_url, client_id, secret)
}

#[cfg(test)]
mod tests {
    use super::*;
    use mockito::Server;

    #[tokio::test]
    async fn test_verify_success() {
        let mut server = Server::new_async().await;
        let mock = server
            .mock("GET", "/auth/v1/verify?code=valid-code")
            .with_status(200)
            .with_header("content-type", "application/json")
            .with_body(r#"{"success":true,"data":{"access_token":"test-token"}}"#)
            .create_async()
            .await;

        let service = ServiceImpl::new(
            server.url(),
            "test-client-id".to_string(),
            "test-secret".to_string(),
        );

        let result = service.verify("valid-code").await;
        mock.assert_async().await;

        assert!(result.is_ok());
        assert_eq!(result.unwrap(), "test-token");
    }

    #[tokio::test]
    async fn test_verify_failure() {
        let mut server = Server::new_async().await;
        let mock = server
            .mock("GET", "/auth/v1/verify?code=invalid-code")
            .with_status(401)
            .with_header("content-type", "application/json")
            .with_body(r#"{"success":false,"message":"Invalid code"}"#)
            .create_async()
            .await;

        let service = ServiceImpl::new(
            server.url(),
            "test-client-id".to_string(),
            "test-secret".to_string(),
        );

        let result = service.verify("invalid-code").await;
        mock.assert_async().await;

        assert!(result.is_err());
    }

    #[tokio::test]
    async fn test_me_success() {
        let mut server = Server::new_async().await;
        let mock = server
            .mock("GET", "/me/v1/me")
            .with_status(200)
            .with_header("content-type", "application/json")
            .with_body(r#"{"success":true,"data":{"id":"user-id","project_id":"","name":"Test User","email":"test@example.com","phone":"","enabled":true,"profile_pic":"","expiry":null,"roles":{},"resources":{},"policies":{},"created_at":null,"created_by":"","updated_at":null,"updated_by":""}}"#)
            .create_async()
            .await;

        let service = ServiceImpl::new(
            server.url(),
            "test-client-id".to_string(),
            "test-secret".to_string(),
        );

        let result = service.me("valid-token").await;
        mock.assert_async().await;

        assert!(result.is_ok());
        let user = result.unwrap();
        assert_eq!(user.id, "user-id");
        assert_eq!(user.name, "Test User");
        assert_eq!(user.email, "test@example.com");
    }

    #[tokio::test]
    async fn test_me_failure() {
        let mut server = Server::new_async().await;
        let mock = server
            .mock("GET", "/me/v1/me")
            .with_status(401)
            .with_header("content-type", "application/json")
            .with_body(r#"{"success":false,"message":"Invalid token"}"#)
            .create_async()
            .await;

        let service = ServiceImpl::new(
            server.url(),
            "test-client-id".to_string(),
            "test-secret".to_string(),
        );

        let result = service.me("invalid-token").await;
        mock.assert_async().await;

        assert!(result.is_err());
    }

    #[tokio::test]
    async fn test_create_resource_success() {
        let mut server = Server::new_async().await;
        let mock = server
            .mock("POST", "/resource/v1/")
            .with_status(200)
            .with_header("content-type", "application/json")
            .with_body(r#"{"success":true,"message":"Resource created successfully"}"#)
            .create_async()
            .await;

        let service = ServiceImpl::new(
            server.url(),
            "test-client-id".to_string(),
            "test-secret".to_string(),
        );

        let resource = Resource::new("resource-1".to_string(), "Test Resource".to_string(), "test-key".to_string());
        let result = service.create_resource(&resource, "valid-token").await;
        mock.assert_async().await;

        assert!(result.is_ok());
    }

    #[tokio::test]
    async fn test_create_resource_failure() {
        let mut server = Server::new_async().await;
        let mock = server
            .mock("POST", "/resource/v1/")
            .with_status(401)
            .with_header("content-type", "application/json")
            .with_body(r#"{"success":false,"message":"Invalid token"}"#)
            .create_async()
            .await;

        let service = ServiceImpl::new(
            server.url(),
            "test-client-id".to_string(),
            "test-secret".to_string(),
        );

        let resource = Resource::new("resource-1".to_string(), "Test Resource".to_string(), "test-key".to_string());
        let result = service.create_resource(&resource, "invalid-token").await;
        mock.assert_async().await;

        assert!(result.is_err());
    }

    #[tokio::test]
    async fn test_service_base_url_trimming() {
        let service = ServiceImpl::new(
            "https://example.com/".to_string(), // URL with trailing slash
            "client".to_string(),
            "secret".to_string(),
        );

        // The base URL should be trimmed
        assert_eq!(service.base_url, "https://example.com");
    }

    #[tokio::test]
    async fn test_basic_auth_encoding() {
        let service = ServiceImpl::new(
            "https://example.com".to_string(),
            "test_client".to_string(),
            "test_secret".to_string(),
        );

        let auth_header = service.basic_auth();
        assert!(auth_header.starts_with("Basic "));

        // Decode and verify the credentials
        let encoded = auth_header.strip_prefix("Basic ").unwrap();
        let decoded = base64::engine::general_purpose::STANDARD
            .decode(encoded)
            .expect("Failed to decode base64");
        let credentials = String::from_utf8(decoded).expect("Failed to convert to string");
        assert_eq!(credentials, "test_client:test_secret");
    }

    #[tokio::test]
    async fn test_verify_network_error() {
        // Test with invalid URL to simulate network error
        let service = ServiceImpl::new(
            "http://invalid-url-that-does-not-exist.local".to_string(),
            "test-client-id".to_string(),
            "test-secret".to_string(),
        );

        let result = service.verify("some-code").await;
        assert!(result.is_err());
        match result.unwrap_err() {
            crate::error::GoIamError::HttpError(_) => {} // Expected
            _ => panic!("Expected HttpError for network failure"),
        }
    }

    #[tokio::test]
    async fn test_verify_malformed_response() {
        let mut server = Server::new_async().await;
        let mock = server
            .mock("GET", "/auth/v1/verify?code=test-code")
            .with_status(200)
            .with_header("content-type", "application/json")
            .with_body("invalid json")
            .create_async()
            .await;

        let service = ServiceImpl::new(
            server.url(),
            "test-client-id".to_string(),
            "test-secret".to_string(),
        );

        let result = service.verify("test-code").await;
        mock.assert_async().await;

        assert!(result.is_err());
        match result.unwrap_err() {
            crate::error::GoIamError::JsonError(_) => {} // Expected
            crate::error::GoIamError::HttpError(_) => {} // Also acceptable - reqwest might handle invalid JSON differently
            _ => panic!("Expected JsonError or HttpError for malformed JSON"),
        }
    }

    #[tokio::test]
    async fn test_verify_success_no_data() {
        let mut server = Server::new_async().await;
        let mock = server
            .mock("GET", "/auth/v1/verify?code=test-code")
            .with_status(200)
            .with_header("content-type", "application/json")
            .with_body(r#"{"success":true}"#)
            .create_async()
            .await;

        let service = ServiceImpl::new(
            server.url(),
            "test-client-id".to_string(),
            "test-secret".to_string(),
        );

        let result = service.verify("test-code").await;
        mock.assert_async().await;

        assert!(result.is_err());
        match result.unwrap_err() {
            crate::error::GoIamError::InvalidResponse { message } => {
                assert!(message.contains("No access token received"));
            }
            _ => panic!("Expected InvalidResponse error"),
        }
    }
}

File: golang/service_impl.go
package golang

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
)

type serviceImpl struct {
	baseURL  string
	clientID string
	secret   string
}

// NewService creates a new instance of the service with the provided base URL, client ID, and secret.
// It returns a Service interface that can be used to interact with the API.
func NewService(baseURL, clientID, secret string) Service {
	return &serviceImpl{
		baseURL:  baseURL,
		clientID: clientID,
		secret:   secret,
	}
}

// Verify sends a verification request with the provided code and returns the access token if successful.
func (s *serviceImpl) Verify(ctx context.Context, code string) (string, error) {
	url := fmt.Sprintf("%s/auth/v1/verify?code=%s", s.baseURL, code)
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return "", fmt.Errorf("error creating request: %w", err)
	}

	req.SetBasicAuth(s.clientID, s.secret)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("error making request: %w", err)
	}
	defer resp.Body.Close()

	var statusError error
	if resp.StatusCode != http.StatusOK {
		statusError = fmt.Errorf("failed to verify code: %s", resp.Status)
	}

	result := AuthCallbackResponse{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		if statusError != nil {
			return "", fmt.Errorf("%w: %s", statusError, err)
		}
		return "", fmt.Errorf("error decoding response: %w", err)
	}

	if !result.Success {
		return "", fmt.Errorf("failed to verify code: %s. Status: %s", result.Message, resp.Status)
	}

	return result.Data.AccessToken, nil
}

// Me retrieves the user information associated with the provided token.
func (s *serviceImpl) Me(ctx context.Context, token string) (*User, error) {
	url := fmt.Sprintf("%s/me/v1/", s.baseURL)
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("error creating request: %w", err)
	}

	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("error making request: %w", err)
	}
	defer resp.Body.Close()

	var statusError error
	if resp.StatusCode != http.StatusOK {
		statusError = fmt.Errorf("failed to fetch user information: %s", resp.Status)
	}

	var user UserResponse
	if err := json.NewDecoder(resp.Body).Decode(&user); err != nil {
		if statusError != nil {
			return nil, fmt.Errorf("%w: %s", statusError, err)
		}
		return nil, fmt.Errorf("error decoding response: %w", err)
	}

	if !user.Success {
		return nil, fmt.Errorf("failed to fetch user information: %s. Status: %s", user.Message, resp.Status)
	}

	return user.Data, nil
}

// ListProjects fetches all projects available to the caller.
func (s *serviceImpl) ListProjects(ctx context.Context, token string) ([]Project, error) {
	url := fmt.Sprintf("%s/project/v1/", s.baseURL)
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("error creating request: %w", err)
	}

	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("error making request: %w", err)
	}
	defer resp.Body.Close()

	var statusError error
	if resp.StatusCode != http.StatusOK {
		statusError = fmt.Errorf("failed to list projects: %s", resp.Status)
	}

	result := ProjectsResponse{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		if statusError != nil {
			return nil, fmt.Errorf("%w: %s", statusError, err)
		}
		return nil, fmt.Errorf("error decoding response: %w", err)
	}

	if !result.Success {
		return nil, fmt.Errorf("failed to list projects: %s. Status: %s", result.Message, resp.Status)
	}

	return result.Data, nil
}

// CreateProject creates a new project with the provided details and token.
func (s *serviceImpl) CreateProject(ctx context.Context, project *Project, token string) error {
	if project == nil {
		return fmt.Errorf("project cannot be nil")
	}

	url := fmt.Sprintf("%s/project/v1/", s.baseURL)
	body, err := json.Marshal(project)
	if err != nil {
		return fmt.Errorf("error marshalling project: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, strings.NewReader(string(body)))
	if err != nil {
		return fmt.Errorf("error creating request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return fmt.Errorf("error making request: %w", err)
	}
	defer resp.Body.Close()

	var statusError error
	if resp.StatusCode != http.StatusOK {
		statusError = fmt.Errorf("failed to create project: %s", resp.Status)
	}

	result := ProjectResponse{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		if statusError != nil {
			return fmt.Errorf("%w: %s", statusError, err)
		}
		return fmt.Errorf("error decoding response: %w", err)
	}
	if !result.Success {
		return fmt.Errorf("failed to create project: %s. Status: %s", result.Message, resp.Status)
	}

	if result.Data != nil {
		*project = *result.Data
	}

	return nil
}

// UpdateProject updates an existing project by ID using the provided details and token.
func (s *serviceImpl) UpdateProject(ctx context.Context, id string, project *Project, token string) error {
	if project == nil {
		return fmt.Errorf("project cannot be nil")
	}

	url := fmt.Sprintf("%s/project/v1/%s", s.baseURL, id)
	body, err := json.Marshal(project)
	if err != nil {
		return fmt.Errorf("error marshalling project: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPut, url, strings.NewReader(string(body)))
	if err != nil {
		return fmt.Errorf("error creating request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return fmt.Errorf("error making request: %w", err)
	}
	defer resp.Body.Close()

	var statusError error
	if resp.StatusCode != http.StatusOK {
		statusError = fmt.Errorf("failed to update project: %s", resp.Status)
	}

	result := ProjectResponse{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		if statusError != nil {
			return fmt.Errorf("%w: %s", statusError, err)
		}
		return fmt.Errorf("error decoding response: %w", err)
	}
	if !result.Success {
		return fmt.Errorf("failed to update project: %s. Status: %s", result.Message, resp.Status)
	}

	if result.Data != nil {
		*project = *result.Data
	}

	return nil
}

// CreateResource creates a new resource with the provided details and token.
// It returns an error if the creation fails. Resource argument will be updated with the created resource details.
func (s *serviceImpl) CreateResource(ctx context.Context, resource *Resource, token string) error {
	url := fmt.Sprintf("%s/resource/v1/", s.baseURL)
	body, err := json.Marshal(resource)
	if err != nil {
		return fmt.Errorf("error marshalling resource: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, strings.NewReader(string(body)))
	if err != nil {
		return fmt.Errorf("error creating request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return fmt.Errorf("error making request: %w", err)
	}
	defer resp.Body.Close()

	var statusError error
	if resp.StatusCode != http.StatusOK {
		statusError = fmt.Errorf("failed to fetch user information: %s", resp.Status)
	}

	result := ResourceResponse{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		if statusError != nil {
			return fmt.Errorf("%w: %s", statusError, err)
		}
		return fmt.Errorf("error decoding response: %w", err)
	}
	if !result.Success {
		return fmt.Errorf("failed to create resource: %s. Status: %s", result.Message, resp.Status)
	}

	return nil
}

// DeleteResource deletes a resource with the provided ID and token.
// It returns an error if the deletion fails.
func (s *serviceImpl) DeleteResource(ctx context.Context, resourceID string, token string) error {
	url := fmt.Sprintf("%s/resource/v1/%s", s.baseURL, resourceID)
	req, err := http.NewRequestWithContext(ctx, http.MethodDelete, url, nil)
	if err != nil {
		return fmt.Errorf("error creating request: %w", err)
	}

	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return fmt.Errorf("error making request: %w", err)
	}
	defer resp.Body.Close()

	var statusError error
	if resp.StatusCode != http.StatusOK {
		statusError = fmt.Errorf("failed to delete resource: %s", resp.Status)
	}

	result := ResourceResponse{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		if statusError != nil {
			return fmt.Errorf("%w: %s", statusError, err)
		}
		return fmt.Errorf("error decoding response: %w", err)
	}

	if !result.Success {
		return fmt.Errorf("failed to delete resource: %s. Status: %s", result.Message, resp.Status)
	}

	return nil
}
